### PR TITLE
fix(source-map-debugger): Consider release files uploaded as indexed archives

### PR DESCRIPTION
**Pull request by lforst**
Fixes an issue with the source map debugger endpoint where we didn't consider release files uploaded as indexed archives.

This PR comes with lots of refactoring.

### CODE REVIEW

File: src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py [hunk 1]
```diff
@@ -238,34 +239,39 @@ def get(self, request: Request, project: Project, event_id: str) -> Response:
...
-    if len(possible_release_files) > 0:
-        path_data["source_file_lookup_result"] = "wrong-dist"
-    for possible_release_file in possible_release_files:
+    artifact_index_release_files = ReleaseFile.objects.filter(
+        organization_id=project.organization_id,
+        release_id=release.id,
+        file__type="release.artifact-index",
+    ).select_related("file")[
+        :1000
+    ]  # limit by something sane in case people have an insane amount of uploads
```
Comment (AbhiPrasad) [inline src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py:256]: m: Do we have metrics that track p95 for release files per release? We can better decide on a magic number if we do.

We should also extract magic number into a constant.

File: src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py [hunk 2]
```diff
@@ -275,143 +281,183 @@ def get_source_file_data(abs_path, project, release, event):
...
-        artifact_bundle__artifactbundleindex__organization_id=project.organization.id,
-        artifact_bundle__artifactbundleindex__url__in=filenme_choices,
-    )
-    if len(possible_release_artifact_bundles) > 0:
-        path_data["source_file_lookup_result"] = (
-            "wrong-dist"
-            if path_data["source_file_lookup_result"] == "unsuccessful"
-            else path_data["source_file_lookup_result"]
+    # Look for source file in artifact indexes
+    if source_file_lookup_result != "found":
```
Comment (AbhiPrasad) [inline src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py:295]: Can we early return instead of having this boolean condition here?

File: src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py [hunk 3]
```diff
@@ -238,34 +239,39 @@ def get(self, request: Request, project: Project, event_id: str) -> Response:
 
 
 def get_source_file_data(abs_path, project, release, event):
-    filenme_choices = ReleaseFile.normalize(abs_path)
+    matching_source_file_names = ReleaseFile.normalize(abs_path)
+    matching_source_map_name = None
+    source_map_reference = None
+    source_file_lookup_result = "unsuccessful"
+    source_map_lookup_result = "unsuccessful"
```
Comment (AbhiPrasad) [inline src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py:246]: m: can we codify lookup states + found source file into a state machine? That way there are only a finite set of possibilities for all of these values - easier to understand what is mutated where.

Can we also type return value? We can use dataclass.

its generally really hard to follow the nested loops/ifs with breaks.

**Review by AbhiPrasad — COMMENTED** [review]
The fact that there is a refactor in this PR makes it really hard to review the specific fix.

File: src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py (hunk 3 above)
Comment (lforst) [reply src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py:246]: Do you mind explaining what you mean by state machine? I mean from a python point of view. How would you implement that?

Comment (AbhiPrasad) [reply src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py:246]: What's hard for me to understand as a reader is all the different permutations of the state variables here. Does `matching_source_map_name` always need to be defined? Or only if `source_map_lookup_result` is `found`?

It also helps that if something is considered `found`, it becomes an invariant to move it from `found` -> something else. that is not clearly defined in the current implementation.

A state machine was just the abstraction I picked that made sense to me, but thinking about it more I can also see a class that has clear helpers and prevents impossible states. You can just call `return source_file_data.to_dict()` or similar to return the expected data from this function.

```py
# has all initial params
matching_source_file_names = ReleaseFile.normalize(abs_path)
source_file_data = SourceFileData(matching_source_file_names)

// ...


```

Comment (lforst) [reply src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py]: Ok I agree the logic was very convoluted in that function. I extracted the whole mechanism into a class now. It's still quite complicated but a bit more readable imo. I think the whole thing is complicated in nature.

Is this more or less what you were looking for?

File: src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py (hunk 2 above)
Comment (lforst) [reply src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py]: I think this might be implicitly resolved.

File: src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py (hunk 1 above)
Comment (lforst) [reply src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py]: Ok, so I talked to @Swatinem and we have to pick a low-ish number here. Each element here will translate into a file download on the endpoint so I picked 25 for now. We can monitor the performance of this endpoint until we LA/GA its feature flag. I don't want to pick something that is _very_ likely to run into issues (i.e. 5) but let's see.

If we don't find it in the 25 results it's not too bad - we will just run into the `unsuccessful` case then and just show "we can't find the source map" in the debugger.

**Comment by lforst** [comment]
> The fact that there is a refactor in this PR makes it really hard to review the specific fix.

@AbhiPrasad I'm sorry about that. I should have refactored first. If you want I can extract this into a separate PR, otherwise I'll know better next time.

File: src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py [hunk 4]
```diff
@@ -237,181 +244,296 @@ def get(self, request: Request, project: Project, event_id: str) -> Response:
...
+        # Cache result
+        if self.artifact_index_release_files is not None:
+            return self.artifact_index_release_files
+
+        self.artifact_index_release_files = ReleaseFile.objects.filter(
+            organization_id=self.project.organization_id,
+            release_id=self.release.id,
+            file__type="release.artifact-index",
+        ).select_related("file")[
+            :25
```
Comment (AbhiPrasad) [inline src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py:515]: l: would like this to be a constant

**Review by AbhiPrasad — APPROVED** [review]
I like the helper class + types, make it way easier to read. Thanks for making the change!
